File: baekjoon/16236.py
from sys import stdin
from collections import deque
from queue import PriorityQueue

dy = [-1, 1, 0, 0]
dx = [0, 0, -1, 1]


# class fish:
#   def __init__(self, y, x, dist, size):
#     self.y = y
#     self.x = x
#     self.dist = dist
#     self.size = size


N = int(stdin.readline())
grid = [[0] * (N + 1) for _ in range(N + 1)]
for i in range(1, N + 1):
  input_list = list(map(int, stdin.readline().split()))
  for j in range(len(input_list)):
    grid[i][j + 1] = int(input_list[j])

shark_y, shark_x = 1, 1
time = 0
shark_size = 2
current_eat = 0
# 먹을만한 물고기들의 예비모임
# 거리,y가 가장 적은것,x가 가장 적은 것 순서대로
# 정렬되어야 하니 우선순위 큐
fish_pool = PriorityQueue()

# 시작점 찾기
for i in range(1, N + 1):
  for j in range(1, N + 1):
    if grid[i][j] == 9:
      shark_y = i
      shark_x = j


def ans():
  global time, shark_size, current_eat, fish_pool, shark_x, shark_y
  while True:
    # 물고기 예비모임 비우기
    while not fish_pool.empty():
      fish_pool.get()

    bfs()

    # 잡아 먹을수 있는 물고기가 아무도 없다면
    if fish_pool.qsize() == 0:
      print(time)
      return

    # 1순위 물고기 잡아먹은후
    # 상어 위치 변경

    grid[shark_y][shark_x] = 0

    # 가장 가깝고
    # 가까운 것들중에서 가장 위쪽에
    # 가깝고 가장 위쪽인것들 중 가장 왼쪽
    dist, y, x = fish_pool.get()

    current_eat += 1
    # 상어의 크기만큼 물고기를 먹었으면
    # 진급해야함
    if current_eat >= shark_size:
      shark_size += 1
      current_eat = 0

    shark_y = y
    shark_x = x
    time += dist

    # 상어가 물고기를 먹고 이동
    grid[shark_y][shark_x] = 9


def bfs():
  global shark_x, shark_y, fish_pool
  check = [[False] * (N + 1) for _ in range(N + 1)]
  q = deque()
  q.append((shark_y, shark_x, 0))
  check[shark_y][shark_x] = True
  while q:
    y, x, dist = q.popleft()
    for k in range(4):
      ny = y + dy[k]
      nx = x + dx[k]

      if ny > N or nx > N or ny < 1 or nx < 1:
        continue

      # 물고기를 탐색
      # 빈 공간(0)이거나 나보다 사이즈가 작거나 같으면
      # 그 공간으로 움직일수 있음
      if not check[ny][nx] and grid[ny][nx] <= shark_size:

        # 뚫려있는 빈 공간이 아니고 상어 크기보다 작은 물고기면
        if grid[ny][nx] != 0 and grid[ny][nx] < shark_size:
          fish_pool.put((dist + 1, ny, nx))
          check[ny][nx]=True
        else :
          check[ny][nx] = True
          q.append((ny, nx, dist + 1))


ans()

File: baekjoon/_16236.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.*;

public class _16236 {

    static int[] dy = {-1, 1, 0, 0};
    static int[] dx = {0, 0, -1, 1};

    static int[][] map;
    static int N;

    static int sharkY = 0;
    static int sharkX = 0;

    static int sharkSize=2;
    static int eat=0;

    private static class Fish {

        int y;
        int x;
        int size;
        int dist;

        public Fish(int y, int x, int size, int dist) {
            this.y = y;
            this.x = x;
            this.size = size;
            this.dist = dist;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            Fish fish = (Fish) o;

            if (y != fish.y) {
                return false;
            }
            if (x != fish.x) {
                return false;
            }
            if (size != fish.size) {
                return false;
            }
            return dist == fish.dist;
        }

        @Override
        public int hashCode() {
            int result = y;
            result = 31 * result + x;
            result = 31 * result + size;
            result = 31 * result + dist;
            return result;
        }
    }

    private static class Shark {

        public Shark(int y, int x, int dist,int size) {
            this.y = y;
            this.x = x;
            this.dist = dist;
            this.size=size;
        }

        int y;
        int x;
        int dist;
        int size;

    }

    private static void printBoard(int[][] map) {
        System.out.println();

        for (int i = 0; i < N; i++) {
            StringBuilder builder = new StringBuilder();
            for (int j = 0; j < N; j++) {
                builder.append(map[i][j]);

            }
            System.out.println(builder);
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());

        map = new int[N][N];

        for (int i = 0; i < N; i++) {
            StringTokenizer s = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                map[i][j] = Integer.parseInt(s.nextToken());
                if (map[i][j] == 9) {
                    sharkY = i;
                    sharkX = j;
                }
            }
        }

        int time = 0;

//        printBoard(map);

        while (true) {
            PriorityQueue<Fish> pq = new PriorityQueue<>((o1, o2) -> {
                if (o1.dist > o2.dist) {
                    return 1;
                } else if (o1.dist == o2.dist) {
                    if (o1.y > o2.y) {
                        return 1;
                    } else if (o1.y == o2.y) {
                        if (o1.x >= o2.x) {
                            return 1;
                        } else {
                            return -1;
                        }
                    } else {
                        return -1;
                    }
                } else {
                    return -1;
                }
            });
            Queue<Shark> q = new LinkedList<>();
            boolean visit[][] = new boolean[N][N];

            q.add(new Shark(sharkY, sharkX, 0,sharkSize));
            visit[sharkY][sharkX] = true;

            findFish(pq, q, visit);

            // 먹을 물고기가 없음
            if (pq.isEmpty()) {
                break;
            }

            // 이제 물고기 먹어야함.
            Fish targetFish = pq.poll();

            eatAndMove(targetFish);
//            printBoard(map);
//            System.out.println("targetFish.dist = " + targetFish.dist);
//            System.out.println("sharkSize = " + sharkSize);
//            System.out.println("eat = " + eat);
            time += targetFish.dist;
        }

        System.out.println(time);


    }

    private static void eatAndMove(Fish targetFish) {
        // 일단 배를 채우고
        eat++;
        // 자기 사이즈만큼 먹었다면 승급
        if (eat==sharkSize){
            eat=0;
            sharkSize++;
        }
        // 물고기 자리로 상어가 이동하고
        map[targetFish.y][targetFish.x]=9;
        // 원래 상어자리는 비게 된다
        map[sharkY][sharkX]=0;
        // 상어의 위치 업데이트
        sharkY= targetFish.y;
        sharkX = targetFish.x;

    }



    private static void findFish(PriorityQueue<Fish> pq, Queue<Shark> q, boolean[][] visit) {
        while (!q.isEmpty()) {
            Shark head = q.poll();

            for (int k = 0; k < 4; k++) {
                int ny = head.y + dy[k];
                int nx = head.x + dx[k];

                // 밖이거나, 방문해봤거나,나보다 큰 물고기는 지나갈 수 없음
              if (isOOB(ny, nx) || visit[ny][nx]||map[ny][nx]>head.size) {
                    continue;
                }

                // 나보다 크기가 작은 물고기를 찾았다면
                if (map[ny][nx]!=0 && map[ny][nx]<head.size){
                    pq.add(new Fish(ny, nx, map[ny][nx], head.dist+1));
                    visit[ny][nx] = true;
                    q.add(new Shark(ny, nx, head.dist + 1, head.size));
                    continue;
                }

                // 아예 빈공간이거나, 내 크기랑 같으면, 지나갈수는 있음
               if (map[ny][nx]==0||(map[ny][nx]!=0&&map[ny][nx]==head.size)){
                    visit[ny][nx] = true;
                    q.add(new Shark(ny, nx, head.dist + 1, head.size));
                    continue;
                }


            }


        }
    }

    private static boolean isOOB(int ny, int nx) {
        return ny >= N || ny < 0 || nx >= N || nx < 0;
    }
}
